cpp/src/arrow/util/formatting.h

```diff
@@ -652,5 +654,82 @@ class StringFormatter<MonthDayNanoIntervalType> {
   }
 };
 
+template <>
+class StringFormatter<BaseListScalar> {
+ public:
+  explicit StringFormatter(const DataType* type) : type_(type) {}
+
+  using value_type = BaseListScalar;
+
+  template <typename Appender>
+  Return<Appender> operator()(const BaseListScalar& value, Appender&& append) {
+    std::stringstream ss;
+    ss << type_->ToString() << "[";
+    for (int64_t i = 0; i < value.value->length(); i++) {
+      if (i > 0) ss << ", ";
+      auto item = value.value->GetScalar(i);
```

Comment:
We shouldn't allocate one scalar per entry of an array in a loop. The virtual-dispatching string formatters should have an initialization phase and then you ask it (via a virtual method) to convert value at position `i` and append it into `std::string *out_appender`.

I will sketch this out and open a draft PR, so you can checkout my branch and implement it for all the nested types. Is that cool with you?

Reply:
I think that’s a good idea. That will help me process and complete the related PR. Let’s proceed with that.

Additionally, your code will provide valuable insights.

Thank you for your help.

Reply:
I haven't forgotten this and started preparing the PR. But it's not my sole focus, so it might take a while.

Reply:
Thank you for your attention and effort. Could you please proceed with this at your convenience? Thank you!